### PR TITLE
bpo-34302: Avoid inefficient way to find start point in deque.index

### PR DESCRIPTION
**Pull request by ksg97031**
Replaced the inefficient loop with faster code from deque_item()

https://bugs.python.org/issue34302

### CODE REVIEW

**Comment by ksg97031** [comment]
```sh
$ cat deque_call_index_with_start_and_stop.py
```
```python
import timeit

queue_setup = '''
from collections import deque
q = deque()
start = 10**5
stop = start + 500
for i in range(0, stop):
    q.append(i)
'''

code = '''
index = q.index(30, 1, stop) // start point is close to entry point
assert index == 30
'''
code2 = '''
index = q.index((start >> 1) + 1, start >> 1, stop >> 1) // start point is close to half point
assert index == (start >> 1) + 1
'''
code3 = '''
index = q.index(start + 1, start, stop) // start point is close to endpoint
assert index == start + 1
'''

repeat = 100000
print(timeit.timeit(setup = queue_setup, stmt = code, number = repeat * 20))
print(timeit.timeit(setup = queue_setup, stmt = code2, number = repeat))
print(timeit.timeit(setup = queue_setup, stmt = code3, number = repeat))
```
## BEFORE
```sh
$ ./python_cur.exe deque_call_index_with_start_and_stop.py
2.154346022
2.899595406
5.265440983
```

## AFTER
```sh
$ ./python_ksg.exe deque_call_index_with_start_and_stop.py
2.1457827320000002
0.717190736
1.9934196979999999
```

# What is changed

Deque block size is fixed to BLOCKLEN.
So we can calculate the number of block with index size.
That means we don't need to increase index one by one for find start point from 0.


## Original Code
```c
for (i=0 ; i<start ; i++) {
    index++;
    if (index == BLOCKLEN) {
        b = b->rightlink;
        index = 0;
    }
}

n = stop - i;
```

## Fixed Code
```c
if (start > 0) {
    i = start + deque->leftindex;
    n = (Py_ssize_t)((size_t) i / BLOCKLEN);
    index = (Py_ssize_t)((size_t) i % BLOCKLEN);

    if (index < (Py_SIZE(deque) >> 1)) {
        while (--n >= 0)
            b = b->rightlink;
    } else {  
        n = (Py_ssize_t)(
                ((size_t)(deque->leftindex + Py_SIZE(deque) - 1))
                / BLOCKLEN - n);
        b = deque->rightblock;
        while (--n >= 0)
            b = b->leftlink;
    }
}

n = stop - start;
```

File: Modules/_collectionsmodule.c [hunk 1]
```diff
@@ -1041,16 +1041,25 @@ deque_index(dequeobject *deque, PyObject *const *args, Py_ssize_t nargs)
...
-        if (index == BLOCKLEN) {
-            b = b->rightlink;
-            index = 0;
+    if (start > 0) {
+        i = start + deque->leftindex;
+        n = (Py_ssize_t)((size_t) i / BLOCKLEN);
+        index = (Py_ssize_t)((size_t) i % BLOCKLEN);
+
+        if (index < (Py_SIZE(deque) >> 1)) {
+            while (--n >= 0)
```
Comment (sir-sigurd) [inline Modules/_collectionsmodule.c:1050]: `braces are required everywhere, even where C permits them to be omitted ...  All new C code requires braces.` ([PEP 7](https://www.python.org/dev/peps/pep-0007/))

File: Modules/_collectionsmodule.c [hunk 2]
```diff
@@ -1041,16 +1041,25 @@ deque_index(dequeobject *deque, PyObject *const *args, Py_ssize_t nargs)
...
+                b = b->rightlink;
+        } else {
+            n = (Py_ssize_t)(
+                    ((size_t)(deque->leftindex + Py_SIZE(deque) - 1))
+                    / BLOCKLEN - n);
+            b = deque->rightblock;
+            while (--n >= 0)
```
Comment (sir-sigurd) [inline Modules/_collectionsmodule.c:1057]: `braces are required everywhere, even where C permits them to be omitted ...  All new C code requires braces.` ([PEP 7](https://www.python.org/dev/peps/pep-0007/))

Comment (ksg97031) [reply Modules/_collectionsmodule.c:1057]: thanks!

**Pull request closed without merging**